### PR TITLE
use dataloader to solve n+1 problem for calenderEventInviteStatus

### LINKED ISSUE #88: [Performance] Fix N+1 query issue in event.inviteStatus
To improve the performance of the query by reducing the no. of separate db queries. Need to batch the event invite status check queries in [event.inviteStatus](https://github.com/coding-blocks/gondor/blob/master/api/graphql/query/CalendarEvent/inviteStatus.js). 

Problem: [N+1](https://engineering.shopify.com/blogs/engineering/solving-the-n-1-problem-for-graphql-through-batching)
Solution: [Dataloader](https://www.npmjs.com/package/dataloader)
Learn: https://www.youtube.com/watch?v=OQTnXNCDywA

### PR DESCRIPTION
**Pull request by prabalsingh24**
fixes #88

### CODE REVIEW

**Comment by 0xnmn** [comment]
@prabalsingh24 It defeats the purpose of dataloader when you are still making N calls for loading N items. 

Try to calculate how it reduces the number of db queries. Right now it doesn't.

You should be making 1 request to find invites for the viewer and many events.

name the loader: `viewerCalendarEventInviteStatus`.
key struct: `{ viewer, event_id }`
loader:
```javascript
new DateLoader(async keys => 
  Models.CalendarEventInvite.findAll({
    where: {
      user_id: keys[0].viewer.id,
      event_id: {
        [Models.Sequelize.Op.in]: keys.map(({ event_id }) => event_id),
      },
    }
  }).then(({ status }) => status)
);
```

The problem left for you to solve is to check/make sure that the order of items returned by the findAll query is same as the keys array otherwise the results will be incorrect.

**Comment by prabalsingh24** [comment]
Okayy. I guess something similar to this was supposed to be done **[user-availability-loader](https://github.com/coding-blocks/gondor/blob/8b355032672d44d0145c59bbdabefc10fa87eff2/api/services/User/index.js#L45)** too

**Comment by 0xnmn** [comment]
> Okayy. I guess something similar to this was supposed to be done **[user-availability-loader](https://github.com/coding-blocks/gondor/blob/8b355032672d44d0145c59bbdabefc10fa87eff2/api/services/User/index.js#L45)** too

Oh, I must have missed it there. Try to solve this one and then work on the previous one again.

**Comment by prabalsingh24** [comment]
@hereisnaman What do you think of this?

```
return Models.CalendarEventInvite.findAll({
          where: {
            user_id: keys[0].user_id,
            event_id: {
              [Models.Sequelize.Op.in]: keys.map(({event_id}) => event_id),
            }
          }
      }).then((invites) => keys.map(key => invites.find(invite => invite.event_id === key.event_id)))
```

This is working. I did not use destruct because there is `invite.event_id` and `key.event_id` and it was getting confusing

**Comment by 0xnmn** [comment]
> @hereisnaman What do you think of this?
> 
> ```
> return Models.CalendarEventInvite.findAll({
>           where: {
>             user_id: keys[0].user_id,
>             event_id: {
>               [Models.Sequelize.Op.in]: keys.map(({event_id}) => event_id),
>             }
>           }
>       }).then((invites) => keys.map(key => invites.find(invite => invite.event_id === key.event_id)))
> ```
> 
> This is working. I did not use destruct because there is `invite.event_id` and `key.event_id` and it was getting confusing

This looks good. Take care of the renaming and that we need to return the status at the end.

File: api/routes/graphql/index.js [hunk 1]
```diff
@@ -3,13 +3,17 @@ import connect from 'Utils/connect';
...
+import CalendarEventInvite from 'Services/CalendarEventInvite';
 import User from 'Services/User';
 
 const server = new ApolloServer({
   schema,
   context: ({ req }) => ({
     viewer: req.viewer,
-    loaders: { userAvailabilityLoader: User.getAvailabilityLoader() },
+    loaders: {
+      userAvailabilityLoader: User.getAvailabilityLoader(),
```
Comment (prabalsingh24) [inline api/routes/graphql/index.js]: I think this `userAvailabilityLoader` --> `userAvailability` should be done. Then we can do `ctx.loaders.userAvailability.load(key)`

**Comment by prabalsingh24** [comment]
> This looks good. Take care of the renaming and that we need to return the status at the end.

Have a look :)

File: api/routes/graphql/index.js (hunk 1 above)
Comment (0xnmn) [reply api/routes/graphql/index.js]: makes sense, go ahead refactor it.

File: api/routes/graphql/index.js [hunk 2]
```diff
@@ -3,13 +3,17 @@ import connect from 'Utils/connect';
...
+      viewerCalendarEventInviteStatus: CalendarEventInvite.getViewerCalendarEventInviteLoader(),
```
Comment (0xnmn) [inline api/routes/graphql/index.js]: rename to: `CalendarEventInvite.getViewerCalendarEventInviteStatusLoader()`

**Review by 0xnmn — CHANGES_REQUESTED** [review]

**Comment by 0xnmn** [comment]
Do the minor changes and lets merge this. Next work on fixing the previous PR.

File: api/services/CalendarEventInvite/index.js [hunk 3]
```diff
@@ -56,4 +57,25 @@ export default class CalendarEventInvite extends BaseModelService {
...
+          [Models.Sequelize.Op.in]: keys.map(({ event_id }) => event_id),
+        },
+      },
+    })
+      .then((invites) =>
+        keys.map((key) =>
+          invites.find((invite) => invite.event_id === key.event_id),
+        ),
+      )
+      .map(({ status }) => status);
```
Comment (0xnmn) [inline api/services/CalendarEventInvite/index.js]: No need to loop over again.

```javascript
 invites.find((invite) => invite.event_id === key.event_id)?.status,
```

**Review by 0xnmn — CHANGES_REQUESTED** [review]

**Comment by prabalsingh24** [comment]
@hereisnaman A final review maybe?

**Comment by prabalsingh24** [comment]
I am getting this bug when I try to decline an event. i don't think it's because of dataloader. I checked in my other branches too. Can you also confirm that this bug isn't due to dataloader?

Screenshot from 2020-05-22 18-52-10

**Review by 0xnmn — APPROVED** [review]

**Pull request merged**